File: 2022/day25/python/solution.py
#!/usr/bin/env python3
import os

def snafu_to_decimal(s):
    """Convert SNAFU number to decimal."""
    digit_values = {'2': 2, '1': 1, '0': 0, '-': -1, '=': -2}
    result = 0
    for char in s:
        result = result * 5 + digit_values[char]
    return result

def decimal_to_snafu(n):
    """Convert decimal number to SNAFU."""
    if n == 0:
        return '0'

    digits = []
    while n:
        remainder = n % 5
        if remainder <= 2:
            digits.append(str(remainder))
            n //= 5
        elif remainder == 3:
            digits.append('=')
            n = n // 5 + 1
        else:  # remainder == 4
            digits.append('-')
            n = n // 5 + 1

    return ''.join(reversed(digits))

def part1(text):
    """Sum all SNAFU numbers and return result as SNAFU."""
    lines = text.strip().split('\n')
    total = sum(snafu_to_decimal(line) for line in lines)
    return decimal_to_snafu(total)

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    with open(input_file) as f:
        text = f.read()

    print('Part 1:', part1(text))
    print('Part 2: No Part 2 on Day 25!')

if __name__ == '__main__':
    main()

File: 2022/day25/ruby/solution.rb
#!/usr/bin/env ruby

def snafu_to_decimal(s)
  digit_values = { '2' => 2, '1' => 1, '0' => 0, '-' => -1, '=' => -2 }
  result = 0
  s.each_char do |char|
    result = result * 5 + digit_values[char]
  end
  result
end

def decimal_to_snafu(n)
  return '0' if n == 0

  digits = []
  while n != 0
    remainder = n % 5
    if remainder <= 2
      digits << remainder.to_s
      n /= 5
    elsif remainder == 3
      digits << '='
      n = n / 5 + 1
    else # remainder == 4
      digits << '-'
      n = n / 5 + 1
    end
  end

  digits.reverse.join
end

def part1(text)
  lines = text.strip.split("\n")
  total = lines.sum { |line| snafu_to_decimal(line) }
  decimal_to_snafu(total)
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  text = File.read(input_file)

  puts "Part 1: #{part1(text)}"
  puts "Part 2: No Part 2 on Day 25!"
end

main if __FILE__ == $PROGRAM_NAME
